### PR TITLE
Implement LandmarkCostFunction 3D.

### PR DESCRIPTION
**Pull request by pifon2a**

### CODE REVIEW

File: cartographer/mapping_2d/pose_graph/landmark_cost_function.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+#include "cartographer/transform/rigid_transform.h"
+#include "cartographer/transform/transform.h"
+#include "ceres/ceres.h"
+#include "ceres/jet.h"
+
+namespace cartographer {
+namespace mapping_2d {
+namespace pose_graph {
+
+// Cost function measuring weighted error between the observed relative pose
```
Comment (gaschler) [inline cartographer/mapping_2d/pose_graph/landmark_cost_function.h]: ...the weighted error between the observed pose given by the landmark measurement and the linearly interpolated pose...
(same in 3D)

File: cartographer/mapping_2d/pose_graph/landmark_cost_function.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+// given by sensor and relative pose computed from the linearly interpolated
+// pose of the robot at the moment of observation.
+class LandmarkCostFunction {
```
Comment (gaschler) [inline cartographer/mapping_2d/pose_graph/landmark_cost_function.h]: You could add some unit tests that cover corner cases where we know the solution.
The unit test should fail if there was a sign error in the implementation.

File: cartographer/mapping_2d/pose_graph/landmark_cost_function.h [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        prev_node_pose[2] + interpolation_parameter_ *
+                                common::NormalizeAngleDifference(
+                                    next_node_pose[2] - prev_node_pose[2])};
+
+    // TODO(pifon2a): Move functions common for all cost functions outside of
+    // SpaCostFunction scope.
+    const std::array<T, 3> unscaled_error =
+        SpaCostFunction::ComputeUnscaledError(landmark_to_tracking_transform_,
+                                              interpolated_pose, landmark_pose);
+
```
Comment (gaschler) [inline cartographer/mapping_2d/pose_graph/landmark_cost_function.h]: remove blank line

File: cartographer/mapping_2d/pose_graph/landmark_cost_function.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    e[1] = T(translation_weight_) * unscaled_error[1];
+    e[2] = T(rotation_weight_) * unscaled_error[2];
+    return true;
+  }
+
+ private:
+  LandmarkCostFunction(const LandmarkObservation& observation,
+                       common::Time prev_node_time, common::Time next_node_time)
+      : landmark_to_tracking_transform_(
+            transform::Project2D(observation.landmark_to_tracking_transform)),
```
Comment (gaschler) [inline cartographer/mapping_2d/pose_graph/landmark_cost_function.h]: Are you sure you can use Project2D in this frame? Is it gravity-aligned?
When the observer's camera points up and you observe a landmark at distance, its Rigid2D projection should still be zero.

File: cartographer/mapping_3d/pose_graph/landmark_cost_function.h [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        4 /* previous node rotation variables */,
+        3 /* next node translation variables */,
+        4 /* next node rotation variables */,
+        3 /* landmark translation variables */,
+        4 /* landmark rotation variables */>(
+        new LandmarkCostFunction(observation, prev_node_time, next_node_time));
+  }
+
+  template <typename T>
+  bool operator()(const T* const prev_node_rotation,
```
Comment (gaschler) [inline cartographer/mapping_3d/pose_graph/landmark_cost_function.h]: This order does not match the template arguments.

File: cartographer/mapping_3d/pose_graph/landmark_cost_function.h [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            interpolation_parameter_ *
+                (next_node_translation[2] - prev_node_translation[2])};
+
+    const Eigen::Quaternion<T> prev_quaternion(
+        prev_node_rotation[0], prev_node_rotation[1], prev_node_rotation[2],
+        prev_node_rotation[3]);
+    const Eigen::Quaternion<T> next_quaternion(
+        next_node_rotation[0], next_node_rotation[1], next_node_rotation[2],
+        next_node_rotation[3]);
+
```
Comment (gaschler) [inline cartographer/mapping_3d/pose_graph/landmark_cost_function.h]: remove blank line

File: cartographer/mapping_3d/pose_graph/landmark_cost_function.h [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    const Eigen::Quaternion<T> interpolated_quaternion =
+        prev_quaternion.slerp(T(interpolation_parameter_), next_quaternion);
+    const T interpolated_pose_rotation[4] = {
```
Comment (gaschler) [inline cartographer/mapping_3d/pose_graph/landmark_cost_function.h]: if the pose interpolation is commonly used, it could go into a helper function.

File: cartographer/mapping_2d/pose_graph/landmark_cost_function.h (hunk 1 above)
Comment (pifon2a) [reply cartographer/mapping_2d/pose_graph/landmark_cost_function.h]: Done

File: cartographer/mapping_2d/pose_graph/landmark_cost_function.h (hunk 3 above)
Comment (pifon2a) [reply cartographer/mapping_2d/pose_graph/landmark_cost_function.h]: Done

File: cartographer/mapping_3d/pose_graph/landmark_cost_function.h (hunk 6 above)
Comment (pifon2a) [reply cartographer/mapping_3d/pose_graph/landmark_cost_function.h]: Done

File: cartographer/mapping_3d/pose_graph/landmark_cost_function.h (hunk 7 above)
Comment (pifon2a) [reply cartographer/mapping_3d/pose_graph/landmark_cost_function.h]: It is not commonly used yet. However, ComputeUnscaledError/ComputeScaledError should go into some lib. Not in this PR though.

File: cartographer/mapping_2d/pose_graph/landmark_cost_function.h (hunk 2 above)
Comment (pifon2a) [reply cartographer/mapping_2d/pose_graph/landmark_cost_function.h]: Almost there.

File: cartographer/mapping_3d/pose_graph/landmark_cost_function.h (hunk 5 above)
Comment (pifon2a) [reply cartographer/mapping_3d/pose_graph/landmark_cost_function.h]: Thanks!

File: cartographer/mapping_2d/pose_graph/landmark_cost_function.h [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        translation_weight_(observation.translation_weight),
+        rotation_weight_(observation.rotation_weight),
+        interpolation_parameter_(
```
Comment (cschuet) [inline cartographer/mapping_2d/pose_graph/landmark_cost_function.h]: Should we check that this is positive?

File: cartographer/mapping_2d/pose_graph/landmark_cost_function.h (hunk 4 above)
Comment (cschuet) [reply cartographer/mapping_2d/pose_graph/landmark_cost_function.h]: I am also not convinced that this is the right thing to do. Let's discuss at a whiteboard.

File: cartographer/mapping_2d/pose_graph/landmark_cost_function.h (hunk 8 above)
Comment (pifon2a) [reply cartographer/mapping_2d/pose_graph/landmark_cost_function.h]: Interpolation parameter? We can check the range [0,1].

File: cartographer/mapping_2d/pose_graph/landmark_cost_function.h (hunk 4 above)
Comment (pifon2a) [reply cartographer/mapping_2d/pose_graph/landmark_cost_function.h]: Added a TODO, it will go into the next PR. This one is already too big.
